internal/monitor_conf.go

```diff
@@ -104,70 +190,124 @@ func getStartCommandLines(cluster *fdbtypes.FoundationDBCluster, processClass fd
+		if err != nil {
+			return configuration, err
+		}
+		for _, argument := range *podSettings.CustomParameters {
+			sanitizedArgument := "--" + equalPattern.ReplaceAllString(argument, "=")
+			configuration.Arguments = append(configuration.Arguments, monitorapi.Argument{Value: sanitizedArgument})
+		}
+	}
 
 	if cluster.Spec.DataCenter != "" {
-		confLines = append(confLines, fmt.Sprintf("locality_dcid = %s", cluster.Spec.DataCenter))
+		configuration.Arguments = append(configuration.Arguments, monitorapi.Argument{Value: fmt.Sprintf("--locality_dcid=%s", cluster.Spec.DataCenter)})
 	}
 
 	if cluster.Spec.DataHall != "" {
-		confLines = append(confLines, fmt.Sprintf("locality_data_hall = %s", cluster.Spec.DataHall))
+		configuration.Arguments = append(configuration.Arguments, monitorapi.Argument{Value: fmt.Sprintf("--locality_data_hall=%s", cluster.Spec.DataHall)})
 	}
 
-	if cluster.Spec.MainContainer.PeerVerificationRules != "" {
-		confLines = append(confLines, fmt.Sprintf("tls_verify_peers = %s", cluster.Spec.MainContainer.PeerVerificationRules))
-	}
+	return configuration, nil
+}
 
-	if cluster.NeedsExplicitListenAddress() && cluster.Status.HasListenIPsForAllPods {
-		confLines = append(confLines, fmt.Sprintf("listen_address = %s", fdbtypes.ProcessAddressesString(cluster.GetFullAddressList("$FDB_POD_IP", false, processNumber), ",")))
+// buildIPArgument builds an argument that takes an IP address from an environment variable
+func buildIPArgument(parameter string, environmentVariable string, imageType FDBImageType, sampleAddresses []fdbtypes.ProcessAddress) []monitorapi.Argument {
+	var leftIPWrap string
+	var rightIPWrap string
+	if imageType == FDBImageTypeUnified {
+		leftIPWrap = "["
+		rightIPWrap = "]"
+	} else {
+		leftIPWrap = ""
+		rightIPWrap = ""
 	}
+	arguments := []monitorapi.Argument{{Value: fmt.Sprintf("--%s=%s", parameter, leftIPWrap)}}
 
-	podSettings := cluster.GetProcessSettings(processClass)
+	for indexOfAddress, address := range sampleAddresses {
+		if indexOfAddress != 0 {
+			arguments = append(arguments, monitorapi.Argument{Value: fmt.Sprintf(",%s", leftIPWrap)})
+		}
 
-	if podSettings.CustomParameters != nil {
-		confLines = append(confLines, *podSettings.CustomParameters...)
-	}
+		arguments = append(arguments,
+			monitorapi.Argument{ArgumentType: monitorapi.EnvironmentArgumentType, Source: environmentVariable},
+			monitorapi.Argument{Value: fmt.Sprintf("%s:", rightIPWrap)},
+			monitorapi.Argument{ArgumentType: monitorapi.ProcessNumberArgumentType, Offset: address.Port - 2, Multiplier: 2},
+		)
 
-	for index := range confLines {
-		for key, value := range substitutions {
-			confLines[index] = strings.Replace(confLines[index], "$"+key, value, -1)
+		flags := make([]string, 0, len(address.Flags))
+		for flag, set := range address.Flags {
+			if set {
+				flags = append(flags, flag)
+			}
+		}
+
+		sort.Slice(flags, func(i int, j int) bool {
+			return flags[i] < flags[j]
+		})
```

Comment:
I think we can move this piece into a common method for the `ProcessAddress` e.g. `SortedFlags()` and make use of it here and in https://github.com/brownleej/fdb-kubernetes-operator/blob/new-image-type/api/v1beta1/foundationdbcluster_types.go#L1947-L1960.